### PR TITLE
feat: Add support for providing input datatypes in TRTorch

### PR DESCRIPTION
**Pull request by peri044**
# Description

This feature adds support for providing input datatypes explicitly. Used for BERT and other common cases where we would require non-float inputs (eg: integers for passing shapes etc). Handles case 3 in https://github.com/NVIDIA/TRTorch/discussions/412#discussioncomment-660967
Pending : Integrate this into `trtorchc`

Fixes https://github.com/NVIDIA/TRTorch/issues/388

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

# Checklist:

- [x] My code follows the style guidelines of this project (You can use the linters)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by narendasan** [comment]
We also need `trtorchc` to be updated

**Comment by narendasan** [comment]
im also wondering if we should create a unified input type that encapsulates: shape, dtype and format, so something like:

```c++
struct Input/Var/Placeholder (Use the PyTorch term here) {
    shape: InputRange 
    dtype: DataType::kFloat
    format: NCHW
};
```

Then we can have a constructor which accepts a just an input range to provide backwards compatibility, or something like that

**Comment by narendasan** [comment]
Or do you think it is easier to have separate aligned lists? It might be less involved.

**Comment by narendasan** [comment]
@peri044 this should be pretty much feature complete, so i think its ready for review, still needs clean up and docs

File: core/ir/Input.cpp [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+  opt = util::toDims(opt_shape);
+  max = util::toDims(max_shape);
+
+  std::vector<int64_t> dyn_shape;
+  for (size_t i = 0; i < opt_shape.size(); i++) {
+    std::set<uint64_t> dim;
+    dim.insert(min_shape[i]);
+    dim.insert(opt_shape[i]);
+    dim.insert(max_shape[i]);
+    LOG_DEBUG(dim.size());
```
Comment (peri044) [inline core/ir/Input.cpp:164]: This debug statement can be more informative.

File: py/trtorch/_compile_spec.py [hunk 2]
```diff
@@ -237,6 +274,7 @@ def TensorRTCompileSpec(compile_spec: Dict[str, Any]) -> torch.classes.tensorrt.
                             "allow_gpu_fallback": false, # (DLA only) Allow layers unsupported on DLA to run on GPU
                         },
                         "op_precision": torch.half, # Operating precision set to FP16
```
Comment (narendasan) [inline py/trtorch/_compile_spec.py:276]: Enabled precisions

File: py/trtorch/_compiler.py [hunk 3]
```diff
@@ -41,6 +43,7 @@ def compile(module: torch.jit.ScriptModule, compile_spec: Any) -> torch.jit.Scri
                         "allow_gpu_fallback": false, # (DLA only) Allow layers unsupported on DLA to run on GPU
                     },
                     "op_precision": torch.half, # Operating precision set to FP16
+                    "input_dtypes": [torch.float32] # List of datatypes that should be configured for each input. Supported options torch.{float|half|int8|int32|bool}.
```
Comment (narendasan) [inline py/trtorch/_compiler.py:46]: delete

File: py/trtorch/_compiler.py [hunk 4]
```diff
@@ -41,6 +43,7 @@ def compile(module: torch.jit.ScriptModule, compile_spec: Any) -> torch.jit.Scri
                         "allow_gpu_fallback": false, # (DLA only) Allow layers unsupported on DLA to run on GPU
                     },
                     "op_precision": torch.half, # Operating precision set to FP16
```
Comment (narendasan) [inline py/trtorch/_compiler.py:45]: enabled precisions

File: py/trtorch/_compiler.py [hunk 5]
```diff
@@ -106,6 +111,7 @@ def convert_method_to_trt_engine(module: torch.jit.ScriptModule, method_name: st
...
```
Comment (narendasan) [inline py/trtorch/_compiler.py:113]: change

File: core/conversion/conversion.cpp [hunk 6]
```diff
@@ -178,7 +188,7 @@ void AddInputs(
 
   ctx->cfg->addOptimizationProfile(profile);
 #if NV_TENSORRT_MAJOR > 7 || (NV_TENSORRT_MAJOR == 7 && NV_TENSORRT_MINOR >= 1)
```
[the remark this answers is not in the document]
Comment (narendasan) [reply core/conversion/conversion.cpp:188]: we will keep it for now, change it with the TRT 8 update

**Review by peri044 — COMMENTED** [review]
Pending local testing otherwise LGTM with minor comments

File: core/ir/Input.cpp [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+//   input_shape = util::toDims(dyn_shape);
+// }
+
+bool valid_dtype_format_combo(nvinfer1::DataType dtype, nvinfer1::TensorFormat format) {
+  switch (dtype) {
+    case nvinfer1::DataType::kINT8: // Supports just Linear (NCHW)
+      switch (format) {
+        case nvinfer1::TensorFormat::kLINEAR:
+          return true;
+        case nvinfer1::TensorFormat::kHWC:
```
[the remark this answers is not in the document]
Comment (narendasan) [reply core/ir/Input.cpp:14]: to be explicit

[the remark this answers is not in the document]
Comment (narendasan) [reply core/ir/Input.cpp:14]: can just remove the return false

File: cpp/api/src/compile_spec.cpp [hunk 8]
```diff
@@ -6,18 +8,131 @@
 #include "trtorch/trtorch.h"
 
 namespace trtorch {
+
+std::ostream& operator<<(std::ostream& os, const CompileSpec::DataType& dtype) {
+  switch (dtype) {
+    case CompileSpec::DataType::kChar:
+      os << "char";
```
[the remark this answers is not in the document]
Comment (narendasan) [reply cpp/api/src/compile_spec.cpp:15]: wanted to use the same term as pytorch and the api at least for the top level, switches to int8 in the core

File: cpp/api/src/compile_spec.cpp [hunk 9]
```diff
@@ -6,18 +8,131 @@
...
+      break;
+    case CompileSpec::DataType::kHalf:
+      os << "half";
+      break;
+    case CompileSpec::DataType::kInt:
+      os << "int";
```
[the remark this answers is not in the document]
Comment (narendasan) [reply cpp/api/src/compile_spec.cpp:21]: wanted to use the same term as pytorch and the api at least for the top level, switches to int32 in the core

[the remark this answers is not in the document]
Comment (narendasan) [reply cpp/api/src/compile_spec.cpp:21]: these are mostly debugging aids for us, I dont think they are used anywhere in the codebase

**Comment by narendasan** [comment]
@peri044 should be good to go

File: cpp/api/src/compile_spec.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (peri044) [reply cpp/api/src/compile_spec.cpp:21]: yeah I think torch has both torch::kINT32 and torch::kInt or at::Int which are aliases

**Pull request merged**